rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp

```diff
@@ -69,7 +70,9 @@ SequentialWriter::~SequentialWriter()
   // Callbacks likely was created after SequentialWriter object and may point to the already
   // destructed objects.
   callback_manager_.delete_all_callbacks();
-  close();
+  if (storage_) {
+    close();
```

Comment:
Relying on the assumption that `storage_` object indicates that the writer is open or has been closed is incorrect here.
The reason for that is for example we can fail to create a new storage in the `bag_split(..)` method and throw an exception. However, we want ti try to write updated `metadata.yaml` file on the destruction of the `writer` object.

The better way would be to create a separate variable protected by a mutex to indicate if the writer is open or closed.

Reply:
Fixed in the commit: https://github.com/ros2/rosbag2/pull/1599/commits/0bd5003e2cb1ca2235580009ef10b2729838c608